File: clients/java/src/main/java/com/google/i18n/myanmar/ZawgyiUnicodeMarkovModel.java
/* Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.i18n.myanmar;

import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Objects;

/**
 * A Markov model to predict whether a string is more likely Unicode or Zawgyi.
 *
 * <p>Internally, this class maintains two Markov chains, one representing Unicode and the other
 * representing Zawgyi. An input string is evaluated against both chains, and the chain that returns
 * the higher probability becomes the prediction.
 *
 * <p>A string is evaluated as a sequence of transitions between states, including transitions to
 * the edges of the string. For example, the string "ABC" contains 4 transitions: NULL to A, A to B,
 * B to C, and C to NULL.
 *
 * <p>For the purposes of Unicode/Zawgyi detection, all characters are treated as the NULL state
 * except for characters in the Myanmar script or characters in the Unicode whitespace range U+2000
 * through U+200B.
 *
 * <p>This class is tested in the google-myanmar-tools-data package because this implementation does
 * not include any logic for building a new instance other than from a binary data stream.
 */
class ZawgyiUnicodeMarkovModel {

  /** Magic number used to identify this object in byte streams. (Reads in ASCII as "UZMODEL") */
  private static final long BINARY_TAG = 0x555A4D4F44454C20L;

  /** Current serial format version number, used in association with the magic number. */
  private static final int BINARY_VERSION = 1;

  // Standard Myanmar code point range before digits
  private static final int STD_CP0 = '\u1000';
  private static final int STD_CP1 = '\u103F';

  // Standard Myanmar code point range after digits
  private static final int AFT_CP0 = '\u104A';
  private static final int AFT_CP1 = '\u109F';

  // Extended Myanmar code point range A
  private static final int EXA_CP0 = '\uAA60';
  private static final int EXA_CP1 = '\uAA7F';

  // Extended Myanmar code point range B
  private static final int EXB_CP0 = '\uA9E0';
  private static final int EXB_CP1 = '\uA9FF';

  // Unicode space characters
  private static final int SPC_CP0 = '\u2000';
  private static final int SPC_CP1 = '\u200B';

  // Indices into Markov nodes
  private static final short STD_OFFSET = 1;
  private static final short AFT_OFFSET = STD_OFFSET + STD_CP1 - STD_CP0 + 1;
  private static final short EXA_OFFSET = AFT_OFFSET + AFT_CP1 - AFT_CP0 + 1;
  private static final short EXB_OFFSET = EXA_OFFSET + EXA_CP1 - EXA_CP0 + 1;
  private static final short SPC_OFFSET = EXB_OFFSET + EXB_CP1 - EXB_CP0 + 1;
  private static final short NUM_STATES = SPC_OFFSET + SPC_CP1 - SPC_CP0 + 1;

  /**
   * Returns the index of the state in the Markov chain corresponding to the given code point.
   *
   * <p>Code points in the standard Myanmar range, Myanmar Extended A, Myanmar Extended B, and
   * Unicode Whitespace each have a unique state assigned to them. All other code points are mapped
   * to state 0.
   *
   * <p>Package-private so that the builder can use this method.
   *
   * @param cp The code point to convert to a state index.
   * @return The index of the state in the Markov chain. 0 <= state < getSize()
   */
  static int getIndexForCodePoint(int cp) {
    if (STD_CP0 <= cp && cp <= STD_CP1) {
      return cp - STD_CP0 + STD_OFFSET;
    }
    if (AFT_CP0 <= cp && cp <= AFT_CP1) {
      return cp - AFT_CP0 + AFT_OFFSET;
    }
    if (EXA_CP0 <= cp && cp <= EXA_CP1) {
      return cp - EXA_CP0 + EXA_OFFSET;
    }
    if (EXB_CP0 <= cp && cp <= EXB_CP1) {
      return cp - EXB_CP0 + EXB_OFFSET;
    }
    if (SPC_CP0 <= cp && cp <= SPC_CP1) {
      return cp - SPC_CP0 + SPC_OFFSET;
    }
    return 0;
  }

  /** The number of states in the Markov chain. */
  static short getSize() {
    return NUM_STATES;
  }

  final BinaryMarkov classifier;

  /** Internal constructor used by ZawgyiUnicodeMarkovModelBuilder */
  ZawgyiUnicodeMarkovModel(BinaryMarkov classifier) {
    this.classifier = classifier;
  }

  /**
   * Creates an instance from a binary data stream.
   *
   * @throws IOException If there was a problem reading the data.
   */
  public ZawgyiUnicodeMarkovModel(InputStream stream) throws IOException {
    DataInputStream dis = new DataInputStream(stream);
    // Check magic number and serial version number
    long binaryTag = dis.readLong();
    if (binaryTag != BINARY_TAG) {
      throw new IOException(
          String.format(
              "Unexpected magic number; expected %016X but got %016X", BINARY_TAG, binaryTag));
    }
    int binaryVersion = dis.readInt();
    if (binaryVersion != BINARY_VERSION) {
      throw new IOException(
          String.format(
              "Unexpected serial version number; expected %08X but got %08X",
              BINARY_VERSION, binaryVersion));
    }
    classifier = new BinaryMarkov(stream);
  }

  /** @param verbose Whether to print the log probabilities for debugging. */
  double predict(String input, boolean verbose) {
    if (verbose) {
      System.out.format("Running detector on string: %s%n", input);
    }

    // Start at the base state
    int prevCp = 0;
    int prevState = 0;

    double totalDelta = 0.0;
    boolean seenTransition = false;
    for (int offset = 0; offset <= input.length(); ) {
      int cp;
      int currState;
      if (offset == input.length()) {
        cp = 0;
        currState = 0;
      } else {
        cp = input.codePointAt(offset);
        currState = getIndexForCodePoint(cp);
      }
      // Ignore 0-to-0 transitions
      if (prevState != 0 || currState != 0) {
        float delta = classifier.getLogProbabilityDifference(prevState, currState);
        if (verbose) {
          System.out.format("U+%04X -> U+%04X: delta=%8.4f ", prevCp, cp, delta);
          for (int i = 1; i < Math.abs(delta); i++) {
            System.out.print("!");
          }
          System.out.println();
        }
        totalDelta += delta;
        seenTransition = true;
      }
      offset += Character.charCount(cp);
      prevCp = cp;
      prevState = currState;
    }

    if (verbose) {
      System.out.format("Final: delta=%.4f%n", totalDelta);
    }

    // Special case: if there is no signal, return -Infinity,
    // which will get interpreted by users as strong Unicode.
    // This happens when the input string contains no Myanmar-range code points.
    if (!seenTransition) {
      return Double.NEGATIVE_INFINITY;
    }

    // result = Pz/(Pu+Pz)
    //        = exp(logPz)/(exp(logPu)+exp(logPz))
    //        = 1/(1+exp(logPu-logPz))
    return 1.0 / (1.0 + Math.exp(totalDelta));
  }

  /**
   * Runs the given input string on both internal Markov chains and computes the probability of the
   * string being unicode or zawgyi.
   *
   * @param input The string to evaluate.
   * @return The probability that the string is Zawgyi given that it is either Unicode or Zawgyi, or
   *     -Infinity if there are no Myanmar range code points in the string.
   */
  public double predict(String input) {
    return predict(input, false);
  }

  @Override
  public boolean equals(Object other) {
    if (!(other instanceof ZawgyiUnicodeMarkovModel)) {
      return false;
    }
    return Objects.equals(classifier, ((ZawgyiUnicodeMarkovModel) other).classifier);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(classifier);
  }
}

File: clients/swift/myanmartools/src/ZawgyiUnicodeMarkovModel.swift
//
//  ZawgyiUnicodeMarkovModel.swift
//  myanmartools
//
//  Created by La Win Ko on 5/6/20.

import Foundation

internal class ZawgyiUnicodeMarkovModel{
    
    // Standard Myanmar code point range before digits
    private let STD_CP0: Character = "\u{1000}"
    private let STD_CP1: Character = "\u{103F}"

    // Standard Myanmar code point range after digits
    private let AFT_CP0: Character = "\u{104A}"
    private let AFT_CP1: Character = "\u{109F}"

    // Extended Myanmar code point range A
    private let EXA_CP0: Character = "\u{AA60}"
    private let EXA_CP1: Character = "\u{AA7F}"

    // Extended Myanmar code point range B
    private let EXB_CP0: Character = "\u{A9E0}"
    private let EXB_CP1: Character = "\u{A9FF}"

    // Unicode space characters
    private let SPC_CP0: Character = "\u{2000}"
    private let SPC_CP1: Character = "\u{200B}"
    
    /** Magic number used to identify this object in byte streams. (Reads in ASCII as "UZMODEL") */
    private let binaryUZMODEL: String = "555A4D4F44454C20"
    private let rangeForbinaryUZMODEL: NSRange = NSRange(location: 0, length: 8)
    
    private let binaryBMARKOV: String = "424D41524B4F5620"
    private var rangeForbinaryBMARKOV: NSRange = NSRange(location: 12, length: 8) //default range unless ssv version is 2
    private let versionOfBMARKOV: Int = 0
    private var rangeForversionOfBMARKOV: NSRange = NSRange(location: 20, length: 4) //default range unless ssv version is 2
    
    private let rangeForSSVTag: NSRange = NSRange(location: 8, length: 4)
    
    private var dataURL: URL
    
    init(dataURL: URL){
        self.dataURL = dataURL
    }
    
    //get parameters
    internal func getParams() -> [[Float32]]{
        
        do{
            if let data = NSData(contentsOf: self.dataURL){
                
                //check UZMODEL
                if binaryUZMODEL != data.getBytes(range: rangeForbinaryUZMODEL)!.hexa{
                    throw NSError(domain: String(format: "Unexpected magic number; expected %@ but got %@", binaryUZMODEL, data.getBytes(range: rangeForbinaryUZMODEL)!.hexa), code: NSFileReadCorruptFileError, userInfo: nil)
                }
                
                var ssv: Int = 0
                
                //check SSV
                ssv = Int(data.getBytes(range: rangeForSSVTag)!.hexa)!
                if ssv == 1{
                    ssv = 0
                }else if ssv == 2{
                    ssv = Int(data.getBytes(range: NSRange(location: 12, length: UInt32.bitWidth))![0])
                }else{
                    throw NSError(domain: String(format: "Unexpected serial version number; expected 1 or 2 but got %i",Int(data.getBytes(range: rangeForSSVTag)![0])), code: NSFileReadUnknownError, userInfo: nil)
                }
                
                //change range start location when SSV version is 2
                if Int(data.getBytes(range: rangeForSSVTag)!.hexa)! != 1{
                    rangeForbinaryBMARKOV = NSRange(location: 16, length: 8)
                }
                
                //check BMARKOV
                if binaryBMARKOV != data.getBytes(range: rangeForbinaryBMARKOV)!.hexa{
                    throw NSError(domain: String(format: "Unexpected magic number; expected %@ but got %@", binaryUZMODEL, data.getBytes(range: rangeForbinaryBMARKOV)!.hexa), code: NSFileReadCorruptFileError, userInfo: nil)
                }
                
                //change range start location when SSV version is 2
                if Int(data.getBytes(range: rangeForSSVTag)!.hexa)! != 1{
                    rangeForversionOfBMARKOV = NSRange(location: 24, length: 4)
                }
                
                //check BMARKOV version
                let tempVersionOfBMARKOV = Int(data.getBytes(range: rangeForversionOfBMARKOV)!.hexa)!
                if tempVersionOfBMARKOV != self.versionOfBMARKOV{
                    throw NSError(domain: String(format: "Unexpected serial version number; expected %i but got %i", self.versionOfBMARKOV, tempVersionOfBMARKOV), code: NSFileReadCorruptFileError, userInfo: nil)
                }
                
                var sizeRange: NSRange
                var size: UInt16 = 0
                
                //change range start location when SSV version is 2
                if Int(data.getBytes(range: rangeForSSVTag)!.hexa)! != 1{
                    sizeRange = NSRange(location: 28, length: 2)
                }else{
                    sizeRange = NSRange(location: 24, length: 2)
                }
                
                //get data size
                let sizeData = NSData(bytes: data.getBytes(range: sizeRange)!, length: 2)
                sizeData.getBytes(&size, length: 2)
                size = UInt16(bigEndian: size)
                
                //read logProbabilityDifferences
                var location = sizeRange.location
                location = location + 2
                var logProbabilityDifferences: [[Float32]] = [[Float32]](repeating: [Float32](repeating: 0, count: Int(size)), count: Int(size))
                
                for i in 0..<size{
                    var entries: Int16 = 0
                    entries = int16Value(bytesArr: data.getBytes(range: NSRange(location: location, length: 2))!)
                    location = location + 2
                    
                    var fallBack: Float32
                    if entries == 0{
                        fallBack = 0
                    }else{
                        fallBack = float32Value(bytesArr: data.getBytes(range: NSRange(location: location, length: 4))!)
                        location = location + 4
                    }

                    var next: Int16 = -1
                    
                    for i2 in 0..<size{
                        if entries > 0 && next < i2{
                            next = int16Value(bytesArr: data.getBytes(range: NSRange(location: location, length: 2))!)
                            location = location + 2
                            entries -= 1
                        }
                        if next == i2{
                            logProbabilityDifferences[Int(i)][Int(i2)] = float32Value(bytesArr: data.getBytes(range: NSRange(location: location, length: 4))!)
                            location = location + 4
                        }else{
                            logProbabilityDifferences[Int(i)][Int(i2)] = fallBack
                        }
                        
                    }
                }
                
                return logProbabilityDifferences

            }else{
                throw NSError(domain: "File not found error", code: NSFileNoSuchFileError, userInfo: nil)
            }
        }catch (let e as NSError){
            print(e.domain)
        }
        
        return [[]]
    }
    
    //mapping myanmar standard unicode indices
    internal func mapping() -> [Character: Int]{
        let beforeDigits: [UInt32] = getChars(start: STD_CP0, end: STD_CP1)
        let afterDigits: [UInt32] = getChars(start: AFT_CP0, end: AFT_CP1)
        let rangeA: [UInt32] = getChars(start: EXA_CP0, end: EXA_CP1)
        let rangeB: [UInt32] = getChars(start: EXB_CP0, end: EXB_CP1)
        let spaceChars: [UInt32] = getChars(start: SPC_CP0, end: SPC_CP1)
        
        //adding all characters into an array
        let myanmarCodePoints: [UInt32] = beforeDigits + afterDigits + rangeA + rangeB + spaceChars
        
        var mapping: [Character: Int] = [:]
        var index = 1
        
        for codePoint in myanmarCodePoints{
            let unicodeScalar = UnicodeScalar(codePoint)
            mapping[Character(unicodeScalar!)] = index
            index += 1
        }
        
        return mapping
    }
    
    /**
     * Runs and get the array of code points between two character range
     *
     * @param start Starting character
     * @param end ending character
     * @return The code points array of UInt32
     */
    internal func getChars(start: Character, end: Character) -> [UInt32]{
        var codePoints: [UInt32] = []
        
        for codePoitnt in start.unicodeScalarCodePoint()...end.unicodeScalarCodePoint(){
            codePoints.append(codePoitnt)
        }
        
        return codePoints
    }
    
    //return float32 value from bytes array
    internal func float32Value(bytesArr: [UInt8]) -> Float32 {
        let data = NSData(bytes: bytesArr, length: 4) as Data
        return Float32(bitPattern: UInt32(bigEndian: data.withUnsafeBytes { $0.load(as: UInt32.self) }))
    }
    
    //return int16 value from bytes array
    internal func int16Value(bytesArr: [UInt8]) -> Int16{
        let data = NSData(bytes: bytesArr, length: 2) as Data
        return Int16(bitPattern: UInt16(bigEndian: data.withUnsafeBytes { $0.load(as: UInt16.self) }))
    }
}
